### ISSUE TITLE
Future of project

### ISSUE THREAD

**Issue by Krzysztof-Cieslak** [ionide/ionide-fsharp#96]
CC: @dsyme @tpetricek @forki @rodrigovidal @cloudRoutine @dmich2 @7sharp9 @rojepp @rneatherway @kjnilsson

Hey guys,

I want to share with You some my thoughts about future of project both in terms of organizational stuff and feature wise. But since it's community-driven project and aimed to help community I don't want to make one-man decisions, but hope for input and help of contributors and all people who pushed me to start it few months ago. 

First of all, few statistics (just for `atom-fsharp`) - 2 months, 310 commits on develop branch, 9 contributors,  2,336 downloads reported by atom page (for comparison `omnisharp-atom` is 9 months old, 864 commits, 20 contributors and 18,627 downloads at the moment). Personally I feel that those are OK-ish numbers for relatively new project and I'm rather happy with community contribution to project ( not only in terms of code contributions but also lot of good opinions on social media and discussions in F# slack channels).

Now, let's move to project features. We have manage to implement following set of features: 
* F# Syntax Highlighting
* Autocomplete
* Error / Warning Highlighting
* Tooltips
* Quick Info Panel
* Symbol Highlighting
* Go-to declaration
* FSI Panel
* Integration with `Paket` (as part of `Paket.Atom`)
* Integration with `FAKE` (should be moved to separate plugin)
* Integration with `generator-fsharp` what allows to scaffold projects from editor.

Yet again I feel that it's OK-ish list of features for 2 months ( but bear in mind that not all features are already perfect ;) ) My short term goals feature wise is adding integration with `Projekt` project (created by @kjnilsson and @rneatherway) to provide better way of handling `fsproj` files. I feel that with `Projekt` integration we could officially release 1.0 version of `atom-fsharp`

1.0 release brings us to more organizational part of equation. I feel we should borrow good ideas from Facebook and theirs http://nuclide.io/ project - provide set of plugins (`atom-fsharp`, `Paket`, `FAKE`, `FsLab` in the future) which can be installed seperatly but also provide common installer for all of them, create some nice webpage for it ( with .io domain as all cool things nowadays ;) ) etc. My suggestion for name would be `IonIDE` - it would be still in same "nuclear physics" style as `Electron`, `Atom` and `Nuclide`. But I'm open for any other suggestions :) 
I'd also want to create `IonIDE` GitHub organization and move there all related projects - `atom-fsharp`, `Paket.Atom`, `generator-fsharp` and future `FAKE.Atom` and `FsLab.Atom`. I have nothing against `fsprojects` organization and I'm very thankful for all support I've got here but I feel it should be incubator (as in its description) - not go-to place for every F# project.   

Last point, I'd like to make `IonIDE` (or whatever we call it) release just before or at DevDay conference in Krakow (16-18 September) - there are few F# talks planned and F# workshop, even though I'm not speaking maybe Tomas would decide to mention it in his talk ;)

Thanks for reading this bit long text and look forward for any opinions and criticism.

**Comment by forki** [comment]
+1 for a separate IonIDE project.

**Comment by GerjanOnline** [comment]
+1 for a separate project, -1 for the name IonIDE (however, I do not have better name, yet)

**Comment by tpetricek** [comment]
First of all, I think the work that you and all the contributors are doing on this is amazing. Getting it from a prototype into something that works and covers pretty much all the essential F# features in just 2 months is pretty epic :-).

I like the idea of focusing on cleaning up everything and making sure that the experience with the current set of features is great. This is a bit boring part, but I think having everything running smoothly is the main thing for 1.0. For example, FsLab integration can probably be added afterwards.

Having a separate organization for the project sounds like a great plan too. I like the `IonIDE` name too, though I would perhaps write it in lower-case as `ionide` with perhaps "ide" in a different color (just to de-emphasize the IDE part a bit - IDEs are so enterprise :-)).

**Comment by 7sharp9** [comment]
Not sure I really get IonIDE unless it's substantially different from atom with re skinning  and lots more base functionality. The face book project adds a lot of new features and changes the whole look as does vscode.  This seems too shallow to even market it as a new project.

**Comment by rneatherway** [comment]
I'm really impressed that things have come along so fast, and it is also very motivational for me to work on FsAutoComplete with more editors using it. It's also very good for F# exposure to have solid support in such a "hip" new editor. Personally I'm not sure if more than a good Github Pages website is necessary, but it certainly can't hurt!

The name is fine, but it I wonder if it might be a bigger sell if it was a bit more of a complete .NET environment. One idea would be to bundle omni-sharp as well and we could look at making Projekt handle C# projects (unless they already have a story for that).

**Comment by 7sharp9** [comment]
Omnisharp is basically Roslyn and non Roslyn support for C# with completion, tips, rename refactor, goto decl, and sone other code nav.

**Comment by Krzysztof-Cieslak** [comment]
@7sharp9 Not sure if I get point about total reskinning - from what I have seen  Nuclide is just set of plugins (for remote debugging, Flow, Hack and Mericurial) with common installer ( actually 2 installers - one Atom based if someone has already installed Atom and just wants plugins, second installing both Atom and plugins) and that's exactly what I want to create.
VS Code is differently at all - it's not Atom based so it just works differently in many places.

Also changing look and feel is in Atom very easy - don't think we should force something particular on users. For example my Atom looks like: 
![image](https://cloud.githubusercontent.com/assets/5427083/8719710/fac7ef10-2bad-11e5-85b9-42b5236409ff.png) - and that's just few plugins and some css

**Comment by 7sharp9** [comment]
Personally I would just work on making projects and solution work correctly the rest is irrelevant really.

**Comment by tpetricek** [comment]
I think branding matters a lot, which is why I think having a catchy name with an IO domain (or not really IO domain, but top-level domain anyway).

It changes the perception from "it also sort of works outside of visual studio" to "there is a full alternative solution". All the stuff on Linux/Mac on fsharp.org helped enormously and this is just another very important step in this direction.

Also, if you are for example data scientist/machine learning person who does not come from a programming background, then having one clear "Download" button is really important.

The Facebook model with one installer for people without Atom and one installer for people with Atom is a very good one IMO.

That said, I think @rneatherway has an interesting thought about supporting Omnisharp too. It might make sense to work with them and see if they'd be interested. Packaging F# support + Omnisharp together and having a strong F# community presence in this would be an obvious win (a lot more visibility for the project, a first-class F# support as part of something bigger). Does anyone know anyone from the Omnisharp project? If no, I can ask around to help get the conversation started.

**Comment by tpetricek** [comment]
Also, some of the things included here - FAKE and Paket - nicely integrate with C# and so having all of these things installed out-of-the-box as one bundle would be great - it would encourage people to look into FAKE and Paket. 

The other thing that people might want would be some unit test runner - not sure if there is anything standard for .NET + Atom.

**Comment by rodrigovidal** [comment]
@Krzysztof-Cieslak You did an amazing job with atom-fsharp.
I didn't know about Projekt, looks very promising. Today the only thing I dislike is the use of FunScript, I'd like to consider moving to TypeScript. 

My name suggestion is "Fermion", it starts with "F" :) (https://en.wikipedia.org/wiki/Fermion)

**Comment by ctaggart** [comment]
I may finally look into https://github.com/Microsoft/TypeScript/issues/3241 to generate TypeScript code from its AST to help the efforts here. Once that is done, a F# wrapper can be made to both read and write TypeScript AST via Edge.js. That would open up the door for FunScript (https://github.com/ZachBray/FunScript/issues/171) and WebSharper to use it. Ideally, I'd like to see this project written in F# and TypeScript. Atom TypeScript is pretty awesome and is written in TypeScript naturally.

**Comment by Krzysztof-Cieslak** [comment]
Bundling also `Omnisharp-atom` is interesting idea indeed, I'll try to ping someone from their team on Twitter. On the other hand I'm bit afraid it will put F# out of center place in project. Also I guess it would require introducing some common standard / conventions for some stuff (for example naming commands)

The fact that `Paket` and `FAKE` works well with C# was one of main reason to put them in separate plugins - i like idea of promoting those tools even more in non-F# community.

I don't think there is any .Net test runner plugin for Atom - creating it might be next interesting thing to write and include in `ionide`

I've also thought about Fermion but I would want to include this "ide" part in name, even if not emphasized a lot, and fermionide sounds weird for me.

I would want to stick with F# - it's good from "marketing" point of view ( creating F# tooling in F#), it makes easier to contribute for people most interested in it (F# developers) and I personally feel it's not so bad right now - IMO, You can contribute without deep understanding of FunScript ( but I guess other contributors can judge it better). We can discuss if we should use FunScript or WebSharper compiler ( no idea how works later) and I guess, it's not really important from point of view of project if we compile to JS or TS (ofc, it would be nice to get as good / fast output code as possible)

**Comment by nosami** [comment]
I've always wanted to be more involved with the F# stuff....... we were planning on using FsAutoComplete ourselves via an omnisharp plugin at some point.... although I'm not sure that there's any benefit in doing that any more - F# already has great editor support.

What I would really like to see is goto definition / find usages etc working in a mixed F#/C# solution... although I'm not sure if that's possible right now.

**Comment by nosami** [comment]
It's probably worth discussing what work is common between the two projects so that we can spread the load. Also, it would be nice to have some consistency in the UI, keybindings etc. 

I also agree with @Krzysztof-Cieslak though - keeping them separate allows us to iterate independently.. although you're always free to bundle omnisharp if you'd like - we don't have any restrictions.

**Comment by david-driscoll** [comment]
@Krzysztof-Cieslak as @nosami said, you're welcome to include it any sort of installed package (that would actually pretty cool imo).

If we need to adjust naming or bindings, just let us know!  I've been meaning to look at F# lately so this might be a good excuse to see what it's like on the other side (will I ever come back? :smile: )

**Comment by 7sharp9** [comment]
I think I agree with @ctaggart.

**Comment by Krzysztof-Cieslak** [comment]
@nosami @david-driscoll Thanks for opinions, guys :) I guess we will think about bundling `omnisharp-atom` in `ionide` (or whatever else we decide to call it). Personally I'm not sure if it will be right choice from start - but it should be added later on if we would together manage to make some improvements in terms of F# / C# mixed solutions. ( But that's just my opinion, I'm not forcing it). 

Also there are some things that could benefit both projects - for example, mentioned here before, .Net test runner which could be created as separate package, so we can bundle it and You could also suggest it for `omnisharp-atom` users

And last thing - after trying F# there is no way back ;)

**Comment by david-driscoll** [comment]
Omnisharp has a feature that returns to us the proper command (mainly for
dnx projects atm) that we use in omnisharp atom.

I had thought about adding more support into Omnisharp, determining what
tests are available, what tests failed (and what have passed or haven't
run).  I just haven't gone down that route yet.  In theory it could be
something that lived outside of omnisharp (and perhaps should).

On Fri, Jul 17, 2015 at 6:25 PM, Krzysztof Cieślak <notifications@github.com

**Comment by Krzysztof-Cieslak** [comment]
Small update:

I'm rather decided to use `ionide` name, and still I would want to release it mid September - so that's bit over 2 weeks left. In past few weeks I was mostly focused on fixing reported bugs and polishing user experience.

I'm also working on webpage (current version: http://ionide.github.io/, logo as it was not accepted by FSSF will be changed - we are working on it), and planning to get ionide.io domain for it.

One thing to be decide is whether to stay with current "small plugins" naming (i.e. atom-fsharp, Paket.Atom) or to use some common pattern for them (for example ionide-fsharp, ionide-paket, ionide-fake) - that's how Facebook's nuclide is doing.

**Comment by nosami** [comment]
Thought I should mention that we have been considering [adding F# support to OmniSharp via a plugin](https://github.com/OmniSharp/omnisharp-roslyn/issues/208) as well as possibly enabling other languages at some point.

I was speaking to one of the VsCode devs a couple of days ago and he would like to use the OmniSharp protocol for F# so that they don't have any future UI integration work (in theory).

**Comment by forki** [comment]
@nosami that is something that should be discussed with the Visual F# team. /cc @NumberByColors @latkin @kevinransom @dsyme

**Comment by Krzysztof-Cieslak** [comment]
@nosami: In my opinion it would be great to have F# Language Service as part of OmniSharp server. But as @forki suggested - those are not my decisions ;) 

If there were F# OmniSharp plugin I would just need to change language services layer of atom-fsharp. Don't see a lot of problems from this point of view.

**Comment by nosami** [comment]
@forki agreed!

@Krzysztof-Cieslak That's great!

**Comment by david-driscoll** [comment]
@nosami so what you're saying is I need to start looking at that plugin
system.... :)

**Comment by Krzysztof-Cieslak** [comment]
[Hello World!] (http://ionide.io/)

OK, So plugin got moved and renamed - from today it's part of `Ionide` organization and is called `ionide-fsharp`. Other plugins (I've removed non-core features and moved them to separate plugins) also have similar names.

I'm also providing `ionide-installer` plugin which is installing all `ionide` plugins and their requirements such as `atom-yeoman` or `linter`.

What's next:
1. Tomas and me have hacked today together proof of concept version of "web preview" panel - it can be used to preview such things like FsLab journal output, FsReveal slides, Suave web applications or any other web application which can be run through FAKE build script. It still needs polishing but it should be ready pretty soon.
2. As more and more people is using editor there some assumptions that are made in several places are not working - I'm thinking about adding special file `.ionide` which would allow to override default conventions we use in `Ionide` on the root folder (solution) level.
3. Test runner - Actually it also should be not as hard to implement as I was thinking - we can run unit tests through FAKE so we have cross platform support and one common API to use out of the box.
